modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/PageMemorySortedIndexStorage.java

```diff
@@ -272,4 +271,97 @@ public R next() {
             }
         };
     }
+
+    private class ScanCursor implements Cursor<IndexRow> {
+        @Nullable
+        private Boolean hasNext;
+
+        @Nullable
+        private final SortedIndexRowKey lower;
+
+        @Nullable
+        private final SortedIndexRowKey upper;
+
+        @Nullable
+        private SortedIndexRow treeRow;
+
+        private ScanCursor(@Nullable SortedIndexRowKey lower, @Nullable SortedIndexRowKey upper) {
+            this.lower = lower;
+            this.upper = upper;
+        }
+
+        @Override
+        public void close() {
+            // No-op.
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (!closeBusyLock.enterBusy()) {
+                throwStorageClosedException();
+            }
+
+            try {
+                advanceIfNeeded();
+
+                return hasNext;
+            } catch (IgniteInternalCheckedException e) {
+                throw new StorageException("Error while advancing the cursor", e);
+            } finally {
+                closeBusyLock.leaveBusy();
+            }
+        }
+
+        @Override
+        public IndexRow next() {
+            if (!closeBusyLock.enterBusy()) {
+                throwStorageClosedException();
+            }
+
+            try {
+                advanceIfNeeded();
+
+                boolean hasNext = this.hasNext;
+
+                this.hasNext = null;
+
+                if (!hasNext) {
+                    throw new NoSuchElementException();
+                }
+
+                return toIndexRowImpl(treeRow);
+            } catch (IgniteInternalCheckedException e) {
+                throw new StorageException("Error while advancing the cursor", e);
+            } finally {
+                closeBusyLock.leaveBusy();
+            }
+        }
+
+        private void advanceIfNeeded() throws IgniteInternalCheckedException {
+            if (hasNext == null) {
```

Comment:
Early return pattern could be used to decrease indentation level for the principal logic of the method

Reply:
Fix it